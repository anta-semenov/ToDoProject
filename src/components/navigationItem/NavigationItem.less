@import '../../styles/variables.less';

.nav-item {
  position: relative;
  box-sizing: border-box;
  padding: 4px 0;
  cursor: pointer;

  &.is-active {    
    cursor: default;
    pointer-events: none;
  }
}

.nav-item__title {
  position: relative;
  font-size: @font-size-basic;
  font-weight: 300;
  color: lighten(@color-font-basic, 40%);
  transition: @transition-basic;

  .nav-item:hover & {
    color: @color-font-basic;
  }
  .nav-item.is-active & {
    font-weight: 500;
    color: @color-font-basic;
  }
}

.nav-item__input {
  position: relative;
  font-size: @font-size-basic;
  font-weight: 300;
  text-decoration: underline;
  color: @color-font-basic;
  border-style: none;
}

.nav-item__count {
  position: relative;
  margin: 0 0 0 8px;

  font-size: @font-size-basic / 1.5;
  font-weight: 400;
  color: @color-font-light;
  vertical-align: top;
}
