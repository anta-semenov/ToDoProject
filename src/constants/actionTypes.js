/*
 * Project
 */
export const ADD_PROJECT = 'ADD_PROJECT'
export const REMOVE_PROJECT = 'REMOVE_PROJECT'
export const COMPLETE_PROJECT = 'COMPLETE_PROJECT'
export const EDIT_PROJECT = 'EDIT_PROJECT'
export const REPLACE_PROJECT = 'REPLACE_PROJECT'

/*
 * Task
 */
export const ADD_TASK = 'ADD_TASK'
export const REMOVE_TASK = 'REMOVE_TASK'
export const COMPLETE_TASK = 'COMPLETE_TASK'
export const EDIT_TASK = 'EDIT_TASK'
export const REPLACE_TASK = 'REPLACE_TASK'
export const ADD_TASK_TO_PROJECT = 'ADD_TASK_TO_PROJECT'
export const ADD_TASK_CONTEXT = 'ADD_TASK_CONTEXT'
export const REMOVE_TASK_CONTEXT = 'REMOVE_TASK_CONTEXT'
export const SET_TASK_TODAY = 'SET_TASK_TODAY'
export const START_TASK_TRACKING = 'START_TASK_TRACKING'
export const STOP_TASK_TRACKING = 'STOP_TASK_TRACKING'

/*
 * Context
 */
export const ADD_CONTEXT = 'ADD_CONTEXT'
export const REMOVE_CONTEXT = 'REMOVE_CONTEXT'
export const EDIT_CONTEXT = 'EDIT_CONTEXT'
export const SWITCH_TASK_CONTEXT = 'SWITCH_TASK_CONTEXT'
export const REPLACE_CONTEXT = 'REPLACE_CONTEXT'

/*
* UI State
*/
export const SET_SELECTED_SECTION = 'SET_SELECTED_SECTION'
export const SET_SIDEBAR_SIZE = 'SET_SIDEBAR_SIZE'
export const SET_ACTIVE_ITEM = 'SET_ACTIVE_ITEM'
export const SET_EDITING_SECTION = 'SET_EDITING_SECTION'
export const TOGGLE_TASK_COMPLETED_LATENCY = 'TOGGLE_TASK_COMPLETED_LATENCY'
export const CLEAR_COMPLETED_LATENT_TASKS = 'CLEAR_COMPLETED_LATENT_TASKS'
export const TOGGLE_TASK_LATENCY = 'TOGGLE_TASK_LATENCY'
export const CLEAR_LATENT_TASKS = 'CLEAR_LATENT_TASKS'
export const SET_SYNCING = 'SET_SYNCING'
export const SET_OFFLINE = 'SET_OFFLINE'
export const SET_AUTH_STATUS = 'SET_AUTH_STATUS'
export const SET_AUTH_ERROR_MESSAGE = 'SET_AUTH_ERROR_MESSAGE'
export const SET_PROPERTY = 'SET_PROPERTY'

/*
* Common
*/
export const SET_STATE = 'SET_STATE'
export const LOG_OUT = 'LOG_OUT'
export const REQUEST_AUTH = 'REQUEST_AUTH'
export const RECIEVE_AUTH = 'RECIEVE_AUTH'
export const ERROR_AUTH = 'ERROR_AUTH'

export const REQUEST_DATA = 'REQUEST_DATA'
export const RECIEVE_DATA = 'RECIEVE_DATA'
export const ERROR_DATA = 'ERROR_DATA'

export const UNDO = 'UNDO'
export const REDO = 'REDO'
export const CLEAR_UNDO_REDO = 'CLEAR_UNDO_REDO'
