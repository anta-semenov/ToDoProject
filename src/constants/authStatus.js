export const AUTH_IN_PROGRESS = 'AUTH_IN_PROGRESS'
export const AUTH_SUCESS = 'AUTH_SUCESS'
export const AUTH_ERROR = 'AUTH_ERROR'
export const AUTH_NONE = 'AUTH_NONE'
