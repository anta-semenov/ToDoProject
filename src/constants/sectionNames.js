export const INBOX = 'Inbox'
export const NEXT = 'Next'
export const TODAY = 'Today'
export const CONTEXTS = 'Contexts'
export const PROJECTS = 'Projects'
export const SOMEDAY = 'Someday'
