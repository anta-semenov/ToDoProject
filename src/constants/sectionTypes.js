export const PROJECTS = 'PROJECTS'
export const CONTEXTS = 'CONTEXTS'
export const PROJECT = 'PROJECT'
export const CONTEXT = 'CONTEXT'
export const INBOX = 'INBOX'
export const TODAY = 'TODAY'
export const NEXT = 'NEXT'
export const SOMEDAY = 'SOMEDAY'
