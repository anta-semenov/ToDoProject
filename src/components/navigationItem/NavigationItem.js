import React from 'react'
import PureRenderMixin from 'react-addons-pure-render-mixin'
import { DropTarget } from 'react-dnd'
import { TODAY, SOMEDAY, PROJECT, CONTEXT } from '../../constants/sectionTypes'
import { TASK } from '../../constants/dndTypes'
import './NavigationItem.less'

const sectionTarget = {
  canDrop: props => props.type === TODAY || props.type === SOMEDAY || props.type === PROJECT || props.type === CONTEXT && !props.editing,
  drop: props => ({
    type: props.type,
    id: props.id
  })
}

const collect = (connect, monitor) => ({
  connectDropTarget: connect.dropTarget(),
  isHovering: monitor.isOver(),
  canDrop: monitor.canDrop()
})

export default class NavigationItem extends React.Component {
  state = {text: ''}
  constructor(props) {
    super(props)
    this.shouldComponentUpdate = PureRenderMixin.shouldComponentUpdate.bind(this)
  }

  componentDidMount() {
    if (this.titleInput) {
      this.titleInput.focus()
    }
  }

  handleKeyDown = (e) => {
    switch (e.keyCode) {
      case 13:
        this.props.onStopEditing({
          type: this.props.type,
          id: this.props.id,
          newTitle: e.target.value
        })
        break
      case 27:
        e.preventDefault()
        this.props.onStopEditing({
          type: this.props.type,
          id: this.props.id
        })
        break
    }
  }

  handleOnBlur = (e) => {
    this.props.onStopEditing({
      type: this.props.type,
      id: this.props.id,
      newTitle: e.target.value
    })
  }

  render() {
    if (this.props.editing) {
      return(
        <li>
          <input
            className='nav-item__input'
            type='text'
            placeholder={this.props.title}
            onKeyDown={this.handleKeyDown}
            onBlur={this.handleOnBlur}
            ref={ref => this.titleInput = ref}
          />
        </li>
      )
    } else {
      return(
        this.props.connectDropTarget(
          <li className={`nav-item ${this.props.active ? 'is-active' : ''} ${this.props.isHovering && this.props.canDrop ? 'nav-item-drop-over' : ''}`} onClick={() => this.props.onItemClick(this.props.type, this.props.id)}>
            <span className='nav-item__title'>{this.props.title}</span>
            {this.props.count ? <span className='nav-item__count'>{this.props.count}</span> : null}
          </li>
        )
      )
    }
  }
}

NavigationItem.propTypes = {
  type: React.PropTypes.string.isRequired,
  title: React.PropTypes.string.isRequired,
  id: React.PropTypes.string,

  onItemClick: React.PropTypes.func.isRequired,
  onStopEditing: React.PropTypes.func,

  active: React.PropTypes.bool.isRequired,
  editing: React.PropTypes.bool,
  count: React.PropTypes.number
}

export const NavigationItemConnectedDropTarget = DropTarget(TASK, sectionTarget, collect)(NavigationItem)
