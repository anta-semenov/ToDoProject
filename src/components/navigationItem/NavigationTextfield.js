import PropTypes from 'prop-types';
import React, { PureComponent } from 'react'
import './NavigationItem.less'

export default class NavigationItem extends PureComponent {
  state = {text: ''}

  componentDidMount() {
    if (this.titleInput) {
      this.titleInput.focus()
    }
  }

  handleKeyDown = (e) => {
    switch (e.keyCode) {
      case 13:
        this.props.onStopEditing({
          type: this.props.type,
          id: this.props.id,
          newTitle: e.target.value
        })
        break
      case 27:
        e.preventDefault()
        this.props.onStopEditing({
          type: this.props.type,
          id: this.props.id
        })
        break
    }
  }

  handleOnBlur = (e) => {
    this.props.onStopEditing({
      type: this.props.type,
      id: this.props.id,
      newTitle: e.target.value
    })
  }

  render() {
    return(
      <li>
        <input
          className='nav-item__input'
          type='text'
          placeholder={this.props.title}
          onKeyDown={this.handleKeyDown}
          onBlur={this.handleOnBlur}
          ref={ref => this.titleInput = ref}
        />
      </li>
    )
  }
}

NavigationItem.propTypes = {
  type: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  id: PropTypes.string,

  onStopEditing: PropTypes.func,

  active: PropTypes.bool.isRequired,
  count: PropTypes.number
}
