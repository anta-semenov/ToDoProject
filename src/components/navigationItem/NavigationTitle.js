import React from 'react'
import shallowCompare from 'react-addons-shallow-compare'
import { DropTarget } from 'react-dnd'
import classNames from 'classnames'
import { TODAY, SOMEDAY, PROJECT, CONTEXT } from '../../constants/sectionTypes'
import { TASK } from '../../constants/dndTypes'
import './NavigationItem.less'

const sectionTarget = {
  canDrop: props => props.type === TODAY || props.type === SOMEDAY || props.type === PROJECT || props.type === CONTEXT && !props.editing,
  drop: props => ({
    type: props.type,
    id: props.id
  })
}

const collect = (connect, monitor) => ({
  connectDropTarget: connect.dropTarget(),
  isHovering: monitor.isOver(),
  canDrop: monitor.canDrop()
})

class NavigationTitle extends React.Component {
  shouldComponentUpdate(nextProps, nextState) {
    return shallowCompare(this, nextProps, nextState)
  }

  render() {
    const navItemClasses = classNames({
      'nav-item': true,
      'is-active': this.props.active,
      'nav-item-drop-over': this.props.isHovering && this.props.canDrop
    })
    
    return this.props.connectDropTarget(
      <li className={navItemClasses} onClick={() => this.props.onItemClick(this.props.type, this.props.id)}>
        <span className='nav-item__title'>{this.props.title}</span>
        {this.props.count ? <span className='nav-item__count'>{this.props.count}</span> : null}
      </li>
    )
  }
}

NavigationTitle.propTypes = {
  type: React.PropTypes.string.isRequired,
  title: React.PropTypes.string.isRequired,
  id: React.PropTypes.string,

  onItemClick: React.PropTypes.func.isRequired,

  active: React.PropTypes.bool.isRequired,
  count: React.PropTypes.number
}

export default DropTarget(TASK, sectionTarget, collect)(NavigationTitle)
