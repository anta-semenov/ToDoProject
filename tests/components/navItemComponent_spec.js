import React from 'react'
import { mount } from 'enzyme'
import * as sectionTypes from '../../src/constants/sectionTypes'
import * as sectionNames from '../../src/constants/sectionNames'

import { NavigationTitleClass } from '../../src/components/navigationItem/NavigationTitle'
import NavigationTextfield from '../../src/components/navigationItem/NavigationTextfield'

const connectDropTarget = element => element
const connectDragSource = element => element
const connectDragPreview = element => element

const testItem1 = {
  key: 0,
  type: sectionTypes.NEXT,
  title: sectionNames.NEXT,
  active: true,
  count: 4,
  editing: false,
  onItemClick: () => {},
  connectDropTarget,
  connectDragSource,
  connectDragPreview
}
const testItem2 = {
  key: 0,
  type: sectionTypes.NEXT,
  title: sectionNames.NEXT,
  active: false,
  editing: false,
  onItemClick: () => {},
  connectDropTarget,
  connectDragSource,
  connectDragPreview
}

describe('Navigation Item', () => {
  describe('Active item class', () => {
    it('Active element should has "is-active" class', () => {
      const itemComponent = mount(<NavigationTitleClass {...testItem1} />)
      expect(itemComponent).toMatchSnapshot()
    })
    it('Inactive element shouldn\'t has "is-active" class', () => {
      const itemComponent = mount(<NavigationTitleClass {...testItem2} />)
      expect(itemComponent).toMatchSnapshot()
    })
  })

  describe('Render not editing component', () => {
    it('Should render component', () => {
      const itemComponent = mount(<NavigationTitleClass {...testItem1} />)
      expect(itemComponent).toMatchSnapshot()
    })
    it('Should invoke callback when clicked with type', () => {
      const onItemClick = jest.fn()
      const testProps = {
        key: 0,
        type: sectionTypes.NEXT,
        title: sectionNames.NEXT,
        active: true,
        count: 4,
        editing: false,
        onItemClick,
        connectDropTarget,
        connectDragSource,
        connectDragPreview
      }

      const itemComponent = mount(<NavigationTitleClass {...testProps} />)
      itemComponent.find('.nav-item').simulate('click')
      expect(onItemClick).toHaveBeenCalledWith(sectionTypes.NEXT, undefined)
    })
    it('Should invoke callback when clicked with type and id', () => {
      const onItemClick = jest.fn()
      const testProps = {
        key: 'PROJECT-bh52ogy5s0fm',
        type: sectionTypes.PROJECT,
        title: 'Some project',
        id: 'bh52ogy5s0fm',
        active: true,
        count: 4,
        editing: false,
        onItemClick,
        connectDropTarget,
        connectDragSource,
        connectDragPreview
      }

      const itemComponent = mount(<NavigationTitleClass {...testProps} />)
      itemComponent.find('.nav-item').simulate('click')
      expect(onItemClick).toHaveBeenCalledWith(sectionTypes.PROJECT, 'bh52ogy5s0fm')
    })
  })

  describe('Render editing component', () => {
    it('Should render editing component', () => {
      const editingProps = {
        key: 'PROJECT-bh52ogy5s0fm',
        type: sectionTypes.PROJECT,
        title: 'Some project',
        id: 'bh52ogy5s0fm',
        active: false,
        count: 4,
        editing: true,
        onItemClick: () => {},
        connectDropTarget,
        connectDragSource,
        connectDragPreview
      }

      const itemComponent = mount(<NavigationTextfield {...editingProps} />)
      expect(itemComponent).toMatchSnapshot()
    })
    it('Should invoke calback with type, and id and new title when press enter', () => {
      const onStopEditing = jest.fn()
      const editingProps = {
        key: 'PROJECT-bh52ogy5s0fm',
        type: sectionTypes.PROJECT,
        title: 'Some project',
        id: 'bh52ogy5s0fm',
        active: false,
        count: 4,
        editing: true,
        onStopEditing,
        connectDropTarget,
        connectDragSource,
        connectDragPreview
      }

      const itemComponent = mount(<NavigationTextfield {...editingProps} />)
      const input = itemComponent.find('.nav-item__input').getDOMNode()
      input.value = 'test'

      itemComponent
        .find('.nav-item__input')
        .simulate('keyDown', { key: 'Enter', keyCode: 13, which: 13 })
      expect(onStopEditing).toHaveBeenCalledWith({
        type: sectionTypes.PROJECT,
        id: 'bh52ogy5s0fm',
        newTitle: 'test'
      })
    })
    it('Should invoke calback with type and id only when press esc', () => {
      const onStopEditing = jest.fn()
      const editingProps = {
        key: 'PROJECT-bh52ogy5s0fm',
        type: sectionTypes.PROJECT,
        title: 'Some project',
        id: 'bh52ogy5s0fm',
        active: false,
        count: 4,
        editing: true,
        onStopEditing,
        connectDropTarget,
        connectDragSource,
        connectDragPreview
      }

      const itemComponent = mount(<NavigationTextfield {...editingProps} />)
      const input = itemComponent.find('.nav-item__input').getDOMNode()
      input.value = 'test'

      itemComponent
        .find('.nav-item__input')
        .simulate('keyDown', { key: 'Esc', keyCode: 27, which: 27 })

      expect(onStopEditing).toHaveBeenCalledWith({
        type: sectionTypes.PROJECT,
        id: 'bh52ogy5s0fm'
      })
    })
  })
})
