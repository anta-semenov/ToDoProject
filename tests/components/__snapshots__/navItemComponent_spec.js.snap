// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation Item Active item class Active element should has "is-active" class 1`] = `
<NavigationTitle
  active={true}
  connectDragPreview={[Function]}
  connectDragSource={[Function]}
  connectDropTarget={[Function]}
  count={4}
  editing={false}
  onItemClick={[Function]}
  title="Next"
  type="next"
>
  <Motion
    defaultStyle={
      Object {
        "translateY": 0,
      }
    }
    style={
      Object {
        "translateY": Object {
          "damping": 26,
          "precision": 0.01,
          "stiffness": 170,
          "val": 0,
        },
      }
    }
  >
    <li
      className="nav-item is-active"
      onClick={[Function]}
      style={
        Object {
          "transform": "translateY(NaNpx)",
        }
      }
    >
      <span
        className="nav-item__title"
      >
        Next
      </span>
      <span
        className="nav-item__count"
      >
        4
      </span>
    </li>
  </Motion>
</NavigationTitle>
`;

exports[`Navigation Item Active item class Inactive element shouldn't has "is-active" class 1`] = `
<NavigationTitle
  active={false}
  connectDragPreview={[Function]}
  connectDragSource={[Function]}
  connectDropTarget={[Function]}
  editing={false}
  onItemClick={[Function]}
  title="Next"
  type="next"
>
  <Motion
    defaultStyle={
      Object {
        "translateY": 0,
      }
    }
    style={
      Object {
        "translateY": Object {
          "damping": 26,
          "precision": 0.01,
          "stiffness": 170,
          "val": 0,
        },
      }
    }
  >
    <li
      className="nav-item"
      onClick={[Function]}
      style={
        Object {
          "transform": "translateY(NaNpx)",
        }
      }
    >
      <span
        className="nav-item__title"
      >
        Next
      </span>
    </li>
  </Motion>
</NavigationTitle>
`;

exports[`Navigation Item Render editing component Should render editing component 1`] = `
<NavigationItem
  active={false}
  connectDragPreview={[Function]}
  connectDragSource={[Function]}
  connectDropTarget={[Function]}
  count={4}
  editing={true}
  id="bh52ogy5s0fm"
  onItemClick={[Function]}
  title="Some project"
  type="PROJECT"
>
  <li>
    <input
      className="nav-item__input"
      onBlur={[Function]}
      onKeyDown={[Function]}
      placeholder="Some project"
      type="text"
    />
  </li>
</NavigationItem>
`;

exports[`Navigation Item Render not editing component Should render component 1`] = `
<NavigationTitle
  active={true}
  connectDragPreview={[Function]}
  connectDragSource={[Function]}
  connectDropTarget={[Function]}
  count={4}
  editing={false}
  onItemClick={[Function]}
  title="Next"
  type="next"
>
  <Motion
    defaultStyle={
      Object {
        "translateY": 0,
      }
    }
    style={
      Object {
        "translateY": Object {
          "damping": 26,
          "precision": 0.01,
          "stiffness": 170,
          "val": 0,
        },
      }
    }
  >
    <li
      className="nav-item is-active"
      onClick={[Function]}
      style={
        Object {
          "transform": "translateY(NaNpx)",
        }
      }
    >
      <span
        className="nav-item__title"
      >
        Next
      </span>
      <span
        className="nav-item__count"
      >
        4
      </span>
    </li>
  </Motion>
</NavigationTitle>
`;
